Implement left_orth/right_orth

mtfishman opened the pull request:
To do:
- [x] Implement block sparse `left_polar`/`right_polar`.
- [x] Add tests.

Future work:
- Implement block sparse `lq_compact`/`lq_full`.

Discussion points:
- Move out-of-place implementation of `left_orth` and `right_orth` to MatrixAlgebraKit.jl?

mtfishman commented:
@lkdvos going into this, I was interested to see how much of the logic in MatrixAlgebraKit.jl could be reused here. The primary issue I found was that `left_orth!` and `right_orth!` in MatrixAlgebraKit.jl are written in a way where they assume the output has been pre-allocated:
- https://github.com/QuantumKitHub/MatrixAlgebraKit.jl/blob/v0.2.0/src/interface/orthnull.jl#L72-L74
- https://github.com/QuantumKitHub/MatrixAlgebraKit.jl/blob/v0.2.0/src/implementations/orthnull.jl#L54-L67

That isn't so straightforward for block sparse matrices and graded arrays since the block sizes and sectors may not be consistent across SVD, polar, and QR/LQ. That meant I had to repeat very similar logic in this PR to circumvent that. This PR makes me think that the MatrixAlgebraKit.jl logic is a bit too opinionated and could match the logic in this PR, I'm curious to hear what you think about that. Maybe MatrixAlgebraKit.jl could have a special codepath for StridedMatrix that allows pre-allocating the output.

lkdvos commented on src/factorizations/orthnull.jl:
```diff
@@ -0,0 +1,96 @@
+  select_algorithm,
+  svd_compact!
+
+function MatrixAlgebraKit.left_orth!(
```
Is this not caught by the MatrixAlgebraKit implementation? If not, it probably should?

mtfishman replied:
It's a bit subtle, the MatrixAlgebraKit.jl version is designed around the output has being pre-allocated, so it has an extra argument for the output. That is tricky to support for block sparse matrices since the different backends (SVD, QR/LQ, polar) might have different block structures/sectors so it isn't straightforward to pre-allocate the output in that way.

We can move this code over to MatrixAlgebraKit.jl, that was the main thing I wanted your feedback on.

mtfishman replied:
Ok, in that case I'll make a PR to MatrixAlgebraKit.jl, it would be nice to not have this logic here.

lkdvos commented on src/factorizations/polar.jl:
```diff
@@ -0,0 +1,53 @@
+
+function MatrixAlgebraKit.left_polar!(A::AbstractBlockSparseMatrix, alg::PolarViaSVD)
+  check_input(left_polar!, A)
+  # TODO: Use more in-place operations here, avoid `copy`.
```
Just checking, the copy is because the multiplication fails right?

lkdvos commented on src/factorizations/polar.jl:
```diff
@@ -0,0 +1,53 @@
+  svd_compact!
+
+function MatrixAlgebraKit.check_input(::typeof(left_polar!), A::AbstractBlockSparseMatrix)
+  @views for I in eachblockstoredindex(A)
+    m, n = size(A[I])
+    m >= n ||
+      throw(ArgumentError("each input matrix block needs at least as many rows as columns"))
+  end
```
In principle this check is also performed when `left_polar` is actually called, so the question is if this error message is more helpful than that one. I'm not saying it isn't, just wanted to pose the question, since I'm fine with either.

lkdvos reviewed (commented):
Overall definitely looks reasonable to me.

As a global comment for the polar decomposition, the question is if we want to implement this as a blockwise polar, or by reimplementing the logic for computing the polar by doing a total svd and then reconstructing the polar decomposition.
Trying to think this through, I'm not sure which solution is better.
On the one hand, blockwise polar decomposition would require again to repeat the logic of constructing the correct axes, but it does seem more in the spirit of the blockwise algorithms.
On the other hand, your current implementation is quite simple and convenient, and probably the efficiency is more or less equal.
Just wanted to hear if you had more thoughts on this?

mtfishman commented:
> As a global comment for the polar decomposition, the question is if we want to implement this as a blockwise polar, or by reimplementing the logic for computing the polar by doing a total svd and then reconstructing the polar decomposition. Trying to think this through, I'm not sure which solution is better. On the one hand, blockwise polar decomposition would require again to repeat the logic of constructing the correct axes, but it does seem more in the spirit of the blockwise algorithms. On the other hand, your current implementation is quite simple and convenient, and probably the efficiency is more or less equal. Just wanted to hear if you had more thoughts on this?

It's a good question. I basically implemented it the current way because it is simpler and leverages the blockwise SVD. I think the leading scaling (in the limit of large blocks) should be the same so I'd lean towards keeping this version for the sake of code simplicity. Maybe we can switch over to the blockwise polar version when we change the logic of all of the blockwise factorizations to be based on first converting to a block diagonal matrix, in which case blockwise factorizations will be simpler to implement.

mtfishman commented:
@lkdvos in the latest, I removed the implementation of polar since I moved it out to #125.

Also, inspired by our discussion above and on Slack, I simplified the logic here by overloading `initialize_output` for block sparse `left_orth!`/`right_orth!` to return `nothing`, so then that can get passed through the existing code logic in MatrixAlgebraKit. Then, I only overload the `left_orth_f!`/`right_orth_f!` functions. It feels slightly hacky but I think it is a reasonable way to go.

mtfishman commented:
Actually, maybe I'll just make that case error for now so we know if we accidentally use that.

lkdvos commented:
I'm not entirely convinced by making that error - if you write generic code that is supposed to handle both dense and blocksparse arrays you would also not be able to provide the output. Realistically it doesn't matter that much yet though, so I'm okay with merging as is

mtfishman commented:
> I'm not entirely convinced by making that error - if you write generic code that is supposed to handle both dense and blocksparse arrays you would also not be able to provide the output. Realistically it doesn't matter that much yet though, so I'm okay with merging as is

I was thinking that too, but also the output isn't easy to construct for block sparse arrays anyway, so how easily can you write generic code across both cases? We can always remove the error if that case comes up, I just figure the error message will make it clear to us when that happens (and generic code can specify the output as `nothing` in the block sparse case anyway).
